core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java

```diff
@@ -0,0 +1,179 @@
+     * @return the payload object
+     * @throws ApplicationException in case of JSON parsing errors or download errors
+     */
+    @SuppressWarnings("unchecked")
+    public Map<String, Object> downloadPayload(String path) {
+        try (InputStream inputStream = externalPayloadStorage.download(path)) {
+            return objectMapper.readValue(IOUtils.toString(inputStream), Map.class);
+        } catch (IOException e) {
+            logger.error("Unable to download payload from external storage path: {}", path, e);
+            throw new ApplicationException(ApplicationException.Code.INTERNAL_ERROR, e);
+        }
+    }
+
+    /**
+     * Verify the payload size and upload to external storage if necessary.
+     *
+     * @param entity      the task or workflow for which the payload is to be verified and uploaded
+     * @param payloadType the {@link PayloadType} of the payload
+     * @param <T>         {@link Task} or {@link Workflow}
+     * @throws ApplicationException       in case of JSON parsing errors or upload errors
+     * @throws TerminateWorkflowException if the payload size is bigger than permissible limit as per {@link Configuration}
+     */
+    public <T> void verifyAndUpload(T entity, PayloadType payloadType) {
+        long threshold = 0L;
+        long maxThreshold = 0L;
+        Map<String, Object> payload = new HashMap<>();
+        String workflowId = "";
+        switch (payloadType) {
+            case TASK_INPUT:
+                threshold = configuration.getTaskInputPayloadSizeThresholdKB();
+                maxThreshold = configuration.getMaxTaskInputPayloadSizeThresholdKB();
+                payload = ((Task) entity).getInputData();
+                workflowId = ((Task) entity).getWorkflowInstanceId();
+                break;
+            case TASK_OUTPUT:
+                threshold = configuration.getTaskOutputPayloadSizeThresholdKB();
+                maxThreshold = configuration.getMaxTaskOutputPayloadSizeThresholdKB();
+                payload = ((Task) entity).getOutputData();
+                workflowId = ((Task) entity).getWorkflowInstanceId();
+                break;
+            case WORKFLOW_INPUT:
+                threshold = configuration.getWorkflowInputPayloadSizeThresholdKB();
+                maxThreshold = configuration.getMaxWorkflowInputPayloadSizeThresholdKB();
+                payload = ((Workflow) entity).getInput();
+                workflowId = ((Workflow) entity).getWorkflowId();
+                break;
+            case WORKFLOW_OUTPUT:
+                threshold = configuration.getWorkflowOutputPayloadSizeThresholdKB();
+                maxThreshold = configuration.getMaxWorkflowOutputPayloadSizeThresholdKB();
+                payload = ((Workflow) entity).getOutput();
+                workflowId = ((Workflow) entity).getWorkflowId();
+                break;
+        }
+
+        try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
```

Comment:
Use `com.google.common.io.CountingOutputStream` to avoid loading entire payload into memory just to get size.

Reply:
The payload needs to be loaded into memory to convert it into an input stream which is used by the upload method.